programs/bpf_loader/src/lib.rs

```diff
@@ -575,89 +670,122 @@ fn process_loader_upgradeable_instruction(
-            })?;
-            programdata.try_account_ref_mut()?.data_as_mut_slice()
-                [programdata_data_offset..programdata_data_offset + buffer_data_len]
-                .copy_from_slice(&buffer.try_account_ref()?.data()[buffer_data_offset..]);
+            {
+                let mut programdata = instruction_context.try_borrow_instruction_account(
+                    transaction_context,
+                    upgradeable_ins_acc_idx::DeployWithMaxDataLen::ProgramData as usize,
+                )?;
+                programdata.set_state(&UpgradeableLoaderState::ProgramData {
+                    slot: clock.slot,
+                    upgrade_authority_address,
+                })?;
+                programdata.get_data_mut()
+                    [programdata_data_offset..programdata_data_offset + buffer_data_len]
+                    .copy_from_slice(&buffer.get_data()[buffer_data_offset..]);
+            }
 
             // Update the Program account
-            program.set_state(&UpgradeableLoaderState::Program {
-                programdata_address: *programdata.unsigned_key(),
-            })?;
-            program.try_account_ref_mut()?.set_executable(true);
+            {
+                let mut program = instruction_context.try_borrow_instruction_account(
+                    transaction_context,
+                    upgradeable_ins_acc_idx::DeployWithMaxDataLen::Program as usize,
+                )?;
+                program.set_state(&UpgradeableLoaderState::Program {
+                    programdata_address,
+                })?;
+                program.set_executable(true);
+            }
 
             if !predrain_buffer {
                 // Drain the Buffer account back to the payer
-                payer
-                    .try_account_ref_mut()?
-                    .checked_add_lamports(buffer.lamports()?)?;
-                buffer.try_account_ref_mut()?.set_lamports(0);
+                let mut payer = instruction_context.try_borrow_instruction_account(
+                    transaction_context,
+                    upgradeable_ins_acc_idx::DeployWithMaxDataLen::Payer as usize,
+                )?;
+                payer.checked_add_lamports(buffer.get_lamports())?;
+                buffer.set_lamports(0);
             }
 
             ic_logger_msg!(log_collector, "Deployed program {:?}", new_program_id);
         }
         UpgradeableLoaderInstruction::Upgrade => {
-            let programdata = keyed_account_at_index(keyed_accounts, first_instruction_account)?;
-            let program = keyed_account_at_index(keyed_accounts, first_instruction_account + 1)?;
-            let buffer = keyed_account_at_index(keyed_accounts, first_instruction_account + 2)?;
-            let rent = get_sysvar_with_account_check::rent(
-                keyed_account_at_index(keyed_accounts, first_instruction_account + 4)?,
+            instruction_context.check_number_of_instruction_accounts(1)?;
```

Comment:
4?

Reply:
4 seems clearer since that is the number of accounts we expect

Reply:
I don't want to mix the "normal" parameters and sysvars as we might disable sysvars being passed as accounts using a feature gate.

Reply:
Its more clear to include all the accounts especially since the sysvars might not always be last.  If we change it the feature gate should include the change in the number of accounts expected

Reply:
> Its more clear to include all the accounts especially since the sysvars might not always be last

Exactly, and combined with the answer to why `check_number_of_instruction_accounts` is called twice this is the reason not to include them. Currently we have to do a separate check for every consecutive run of "normal" accounts until it is interrupted by a sysvar. But once we feaure gate the sysvar account check, we can merge all these `check_number_of_instruction_accounts` into one which covers all account indices at once.